ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,284 @@
+package ring
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	totalTokensCount         = 1 << 32
+	optimalTokensPerInstance = 1 << 9
+	maxZonesCount            = 1 << 3
```

Comment:
[nit] Do we really need bit wise operators here? Can't we just put the actual number to help anyone else different than chatgpt to read it? 😉